Tiber upgrade (#259)
autonity/docs.autonity.org: concepts/architecture/index.md

Reviewer: Extensive and well-written changes here!

- Protocol smart contracts: the Autonity protocol is deployed in part via protocol smart contracts:

	- **Autonity Protocol Contract** implementing protocol primitives for governance, tokenomics,  liquid staking, block finalisation, and rewards distribution.
	- **Liquid Newton** contracts for validator-specific liquid stake tokens.

> 	- **Accountability Contracts** implementing protocol primitives for accountable and omission fault detection for validators participating in the consensus committee. Accountability covers participation in consensus and oracle protocols. The consensus accountability protocol enforces validator liveness and adherence to the [Tendermint consensus](/concepts/consensus/pos/) rules, implementing slashing penalties and a [Penalty-Absorbing Stake (PAS)](/concepts/afd/#penalty-absorbing-stake-pas) model.

Reviewer 2: ```suggestion
	- **Accountability Contracts** implementing protocol primitives for accountable and omission fault detection for validators participating in the consensus committee. The consensus accountability protocol enforces validator liveness and adherence to the [Tendermint consensus](/concepts/consensus/pos/) rules, implementing slashing penalties and a [Penalty-Absorbing Stake (PAS)](/concepts/afd/#penalty-absorbing-stake-pas) model.
```

Reviewer 2: Since the next sentence about the oracle protocol also mentions accountability, I think we can omit it here.

	- **Autonity Oracle Contract** implementing protocol primitives for the oracle protocol. It provides functionality for computing aggregated median price data from external price data and managing the set of currency pairs for which Autonity's [oracle network](/concepts/oracle-network/) provides price data. It provides functionality for an oracle accountability protocol incentivising timely participation in oracle voting rounds and the submission of accurate oracle price reports by validators.

> 	- **Auton Stabilization Mechanism Contracts** implementing a [Stabilization Mechanism](/concepts/asm/) for the protocol's native coin [Auton](/concepts/protocol-assets/auton/) via a Collateralized Debt Position or CDP-based stabilization mechanism. Changes in supply and demand for Auton are absorbed by dynamically adjusting CDP incentives to increase and decrease Auton borrowing costs when Auton price moves above or below its Stabilization Target the [Autonomous Currency Unit (ACU)](/glossary/#acu).

Reviewer 2: ```suggestion
	- **Auton Stabilization Mechanism Contracts** implementing a [Stabilization Mechanism](/concepts/asm/) for the protocol's native coin [Auton](/concepts/protocol-assets/auton/) via a Collateralized Debt Position (CDP). Changes in supply and demand for Auton are absorbed by dynamically adjusting CDP incentives to increase and decrease Auton borrowing costs when Auton price moves above or below its Stabilization Target the [Autonomous Currency Unit (ACU)](/glossary/#acu).
```

- A percentage determined by the protocol `treasuryFee` parameter is deducted and transferred to the protocol treasury account for community funds.
- Rewards are distributed to the treasury account of each committee member (validator) on a _pro rata_ basis, depending on their share of the bonded stake in the consensus committee.
- A percentage determined by the validator `commissionRate` parameter is deducted and transferred to the validator treasury account as a commission fee. The initial commission rate is set globally for all validators in the network, specified by the `delegationRate` protocol parameter in the genesis configuration file.  Individual validators can modify their commission rate after registration.
- Rewards are distributed to stake delegator accounts _pro rata_ according to their share of the stake bonded to the validator.
- Rewards accumulate until claimed by stake delegators. 

To learn more about the concept see [Staking rewards and distribution](/concepts/staking/#staking-rewards-and-distribution) and [Staking accounts](/concepts/staking/#staking-accounts).

> Inflation rewards accumulate from the Newton inflation mechanism and awarded for Newton stake token that is delegated to validators. The emissions are:

Reviewer 2: something missing here

Author: Yes, this should be re-written to:

>Inflation rewards accumulate from the Newton inflation mechanism and are awarded for Newton stake token that has been delegated to validators in the consensus committee. The emissions are:

Author: Done in commit: https://github.com/autonity/docs.autonity.org/commit/0331489abdd49ad0ec1c61a194952ade8ccb73d9.

> - Computed for delegated stake according to the Newton [inflation mechanism](/glossary/#inflation-mechanism) (note that _all_ delegated stake earns inflation rewards - there is no dependency on stake being delegated in the current consensus committee).

Reviewer 2: Also self-bonded stake gets inflation rewards autobonded. Also the part in the "(...)" is incorrect, inflation rewards are given only to stake which is currently in the committee

Author: I agree looking at the code, though I was told otherwise. I will change this so:

>Inflation rewards accumulate from the Newton inflation mechanism and are awarded for Newton stake token that has been delegated to validators in the consensus committee. The emissions are:
>
>- Computed for stake bonded to the consensus committee members according to the Newton [inflation mechanism](/glossary/#inflation-mechanism) (note inflation rewards are _only_ earned for stake actively securing the network - stake delegations to validators that _are not_ in the current consensus committee _will not_ earn inflation rewards).

Commit: 0331489abdd49ad0ec1c61a194952ade8ccb73d9

> - The Newton inflation emission is minted from the [inflation reserve](/glossary/#inflation-mechanism) and [automatically bonded](/glossary/#autobond) Block _base fees_ are added to the rewards pool and distributed at the end of the epoch.

Reviewer 2: ```suggestion
- The Newton inflation emission is minted from the [inflation reserve](/glossary/#inflation-mechanism) and [automatically bonded](/glossary/#autobond). Block _base fees_ are added to the rewards pool and distributed at the end of the epoch.
```

To learn more about the concept see [Total supply and newton inflation ](/concepts/protocol-assets/newton/#total-supply-and-newton-inflation) in the [Newton](/concepts/protocol-assets/newton) concept.

### Autonity Accountability Contract
The contract implementing the accountability fault detection (AFD) protocol extensions, including primitives for misbehaviour accusations, proving innocence against an accusation, proven faults, slashing, and jailing.

The contract stores static [protocol configuration](/concepts/afd/#protocol-configuration) used to compute slashing penalties. Contract functions are called by validators whilst participating in the AFD protocol to:

- Return a committee member's proven faults
- Determine if a new accusation can be made and is slashable
- Submit accountability events.

> Function calls to compute accountability each block and apply accountability penalties at epoch end are restricted to protocol.

Reviewer 2: ```suggestion
Function calls to compute accountability at each block and apply accountability penalties at epoch end are restricted to protocol.
```

"at " missing I guess

- Apply slashing according to Autonity's [Penalty-Absorbing Stake (PAS)](/concepts/afd/#penalty-absorbing-stake-pas) model: validator self-bonded stake is slashed first until exhausted, then delegated stake.
- Depending upon fault severity, a slashing penalty may also apply temporary or permanent validator jailing to bar the validator from committee selection. The severity of the jailing period is influenced by the severity of the fault committed and the validator's slashing history over time. See [Jail period calculation](/concepts/afd/#jail-period-calculation).

To learn more about the concept see [Accountability fault detection (AFD)](/concepts/afd/).

### Autonity Omission Accountability Contract
The contract implementing the omission fault detection (OFD) protocol extensions, including primitives for detecting failure to participate in consensus voting rounds, slashing, and jailing.

The contract stores static [protocol configuration](/concepts/ofd/#protocol-configuration) used to compute slashing penalties.

> Function calls to detect validator inactivity each block and apply omission penalties at epoch end are restricted to protocol.

Reviewer 2: ```suggestion
Function calls to detect validator inactivity at each block and apply omission penalties at epoch end are restricted to protocol.
```

All functions are documented in the Reference [Autonity Interfaces](/reference/api/): public API's under [Accountability Contract Interface](/reference/api/afd/), governance under [Governance and Protocol Only Reference](/reference/api/aut/op-prot/).

#### Inactivity scoring

> The Autonity Omission Accountability Contract implements logic for measuring if a consensus committee is actively participating in consensus voting rounds as a committee member:

Reviewer 2: ```suggestion
The Autonity Omission Accountability Protocol implements logic for measuring if a committee member is actively participating in consensus voting rounds:
```

> - Measures validator activity by recording committee member precommit messages as a cryptographically-verifiable BLS aggregate signature as an $ActivityProof$ generated by the block proposer and included in the block headers.

Reviewer 2: ```suggestion
- Measures validator activity by recording precommit messages as a cryptographically-verifiable BLS aggregate signature ($ActivityProof$) generated by the block proposer and included in the block headers.
```

> - Maintains an inactivity score for committee members during the epoch, computed as an offline percentage of blocks in an epoch that the validator has failed to participate in consensus for.

Reviewer 2: ```suggestion
- Maintains an inactivity score for committee members during the epoch, computed as a percentage of blocks in an epoch that the validator has failed to participate in consensus for.
```

#### Slashing penalty computation
The Autonity Omission Accountability Contract manages the computation of omission penalties for proven faults at epoch end. A slashing model is implemented with penalties increasing in severity as inactivity thresholds are passed. The contract implements logic to:

> - Calculate penalties based on static (set in the contract, see [protocol configuration](/concepts/ofd/#protocol-configuration)) and dynamic factors specific to the epoch circumstances. See [slashing amount calculation](/concepts/ofd/#slashing-amount-calculation).

Reviewer 2: I am not sure if it is because this docs is based on the previous release, but currently I believe that all parameters of the omission contract can be changed (i.e. there are no static parameters).

Author: Hmm. It is static as it is set and remains until it is changed, but I see the point. Hmm.

Will remove the static and rely on protocol config. Revised so:

>Calculate penalties based on parameters set in the [protocol configuration](/concepts/ofd/#protocol-configuration) and dynamic factors specific to the epoch circumstances

Done in commit 769e9c24eb1b3ee1bab3627614dd7bbb5ef77662

Author: I reviewed the whole page for 'static' and fixed in the same commit 769e9c24eb1b3ee1bab3627614dd7bbb5ef77662.

Depending upon the validator's inactivity score the protocol will apply penalties to:

- Withhold staking and newton inflation rewards. See [Rewards withholding calculation](/concepts/ofd/#rewards-withholding-calculation).
- Withhold rewards and apply jailing coupled with a probation period. See [Probation period calculation](/concepts/ofd/#probation-period-calculation) and [Jail period calculation](/concepts/ofd/#jail-period-calculation).

> - Apply slashing and permanent jailing if repeat offending while on probation. See [Slashing amount calculation](/concepts/ofd/#slashing-amount-calculation). Slashing is applied according to Autonity's [Penalty-Absorbing Stake (PAS)](/concepts/afd/#penalty-absorbing-stake-pas) model: validator self-bonded stake is slashed first until exhausted, then delegated stake.

Reviewer 2: ```suggestion
- Apply slashing and longer jailing if repeated offences are committed while on probation. See [Slashing amount calculation](/concepts/ofd/#slashing-amount-calculation). Slashing is applied according to Autonity's [Penalty-Absorbing Stake (PAS)](/concepts/afd/#penalty-absorbing-stake-pas) model: validator self-bonded stake is slashed first until exhausted, then delegated stake.
```

Jailbounding is really just an edge case for when we end up slashing 100%

To learn more about the concept see [Omission fault detection (OFD)](/concepts/ofd/).

### Autonity Oracle Contract
The contract implementing the Oracle protocol extensions, including primitives for computing median price data, managing the set of currency pairs for which Autonity provides price data, and oracle accountability fault detection.

The contract stores:

> - [Protocol parameters](/reference/protocol/) for median price computation and oracle voting rounds. These specify the currency pairs for which the oracle provides median price data, the interval over which an oracle round for submitting and voting on price data runs.

Reviewer 2: the price we end up providing is not really a median, but rather a weighted average which uses the confidence as weight. The median is used only for outlier detection.

Author: Yes, need to replace 'median price' with 'weighted average'. Reviewed and corrected throughout the section, median only remains in relation to outlier detection.

See commit 79c5b3485b1bb1e4be717030e29ca7def6af6a58

> - Static [protocol configuration](/concepts/oafd/#protocol-configuration) used to compute slashing penalties.

Reviewer 2: Which part of the config is the static part you are referring to?

Author: See preceding comment thread on static and change made https://github.com/autonity/docs.autonity.org/pull/259#discussion_r2035642861

Fixed in commit 769e9c24eb1b3ee1bab3627614dd7bbb5ef77662.

- Set oracle voters based on validators in the consensus committee and update the voter set as the consensus committee is re-selected at the end of an epoch.
- Manage oracle voting rounds, triggering the initiation of a new voting period at the end of a round.

#### Voter selection
Participation in the oracle protocol is a validator responsibility and validators in the consensus committee are automatically selected to vote on weighted average price computation by a protocol-only function. As the last block of an epoch is finalized, this function is executed to determine the oracle voters for the following epoch.

Consensus committee membership is computed by the Autonity Protocol Contract; see [committee selection](/concepts/architecture/#committee-selection).

#### Slashing penalty computation
The Autonity Oracle Contract manages the computation of accountability penalties for proven oracle price reporting faults at epoch end. A slashing model is implemented with penalties applied for reporting "outlier" prices falling outside a tolerance threshold of the median prices reported by other oracles in the [oracle network](/concepts/oracle-network). The contract implements logic to:

> - Calculate penalties based on parameters set in the [protocol configuration](/concepts/ofd/#protocol-configuration) and dynamic factors specific to the epoch circumstances. See [slashing amount calculation](/concepts/ofd/#slashing-amount-calculation).

Reviewer: > - Calculate penalties...

This is an unnecessary bullet point (solo item does not necessitate a list).

Author: I'm trying to keep consistent formatting style throughout the sections, hence a solo point, but removed in commit c4ccb4dccb62cee53501edb2817c2d888fdb686b and edited to:

>... The contract calculates penalties based on parameters set in the [protocol configuration](/concepts/oafd/#protocol-configuration) and dynamic factors specific to the epoch circumstances. See [slashing amount calculation](/concepts/ofd/#slashing-amount-calculation).
